src/cell_list.h

```diff
@@ -0,0 +1,226 @@
+#pragma once
+#ifndef CELL_LIST_H
+#define CELL_LIST_H
+
+#include <algorithm>
+#include <list>
+#include <vector>
+
+#include "cata_utility.h"
+#include "line.h"
+#include "point.h"
+
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list_iterator;
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list_range;
+
+/*
+ * A Cell List operates by subdividing an area into a regular grid and populating each cell
+ * with a list of the entities in that area.
+ * It is used to optimize lookup of nearby entities.
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list
+{
+    public:
+        cell_list() = default;
+        void add( const point &p, T value ) {
+            std::pair<point, T> val( p, value );
+            cells[p.x / CELL_SIZE][p.y / CELL_SIZE].push_back( val );
+        }
+        cell_list_range< T, CELL_SIZE, MAXX, MAXY > get( const point &p, int distance ) {
+            return cell_list_range<T, CELL_SIZE, MAXX, MAXY>( *this, p, distance );
+        }
+        static constexpr int num_cells_x() {
+            return divide_round_up( MAXX, CELL_SIZE );
+        }
+        static constexpr int num_cells_y() {
+            return divide_round_up( MAXY, CELL_SIZE );
+        }
+        friend class cell_list_range< T, CELL_SIZE, MAXX, MAXY >;
+    private:
+        std::vector<std::pair<point, T>> cells[num_cells_x()][num_cells_y()];
+};
+/*
+ * A cell list range handles iteration over elements in the Cell List near some origin point in
+ * nearest-first order and terminates at some specified threshold.
+ * It does this by visiting one "rank" of cells at a time.
+ * First it visits the cell the origin is in, determining the distance to the nearest cell wall,
+ * and iterating over every entity closer than that distance. This guarantees that there is no
+ * closer entity than one found via this process. Once that list is exhausted, it loads one row of
+ * cells around the origin cell and repeating the above process.
+ */
+template< typename T, int CELL_SIZE, int MAXX, int MAXY >
+class cell_list_range
+{
+    public:
+        cell_list_range( const cell_list< T, CELL_SIZE, MAXX, MAXY > &list, const point &p,
+                         int distance ) : list( &list ), origin( p ), origin_cell( p / CELL_SIZE ),
+            max_distance( distance ), sorted_elements( distance + 1, std::vector<std::pair<point, T >>() ) {
```

Comment:
You presumably won't need all `distance + 1` elements in the most common case, but the constructor creates them all anyways. Wouldn't it be better to add them on demand (optionally only `reserve`ing in the constructor)?

Reply:
I mostly just don't want to complicate things by checking for size and doing push_back() into the vector on every insertion into sorted_elements.  Empty vectors are extremely lightweight, I don't want to complicate this unless profiling shows this as a bottleneck.
Also optimization could go either way because a resize could result in copying an existing sorted_elements vector, which could be quite expensive.